xilem_core/src/sequence.rs

```diff
@@ -1,363 +1,650 @@
-        /// Internally it works by having a pointer/index to the current/old element (0 at the beginning),
-        /// and the pointer is incremented by basically all methods that mutate that sequence.
-        pub trait $elements_splice {
-            /// Insert a new element at the current index in the resulting collection (and increment the index by 1)
-            fn push(&mut self, element: $pod, cx: &mut $cx);
-            /// Mutate the next existing element, and add it to the resulting collection (and increment the index by 1)
-            fn mutate(&mut self, cx: &mut $cx) -> &mut $pod;
-            // TODO(#160) this could also track view id changes (old_id, new_id)
-            /// Mark any changes done by `mutate` on the current element (this doesn't change the index)
-            fn mark(&mut self, changeflags: $changeflags, cx: &mut $cx) -> $changeflags;
-            /// Delete the next n existing elements (this doesn't change the index)
-            fn delete(&mut self, n: usize, cx: &mut $cx);
-            /// Current length of the elements collection
-            fn len(&self) -> usize;
-            // TODO(#160) add a skip method when it is necessary (e.g. relevant for immutable ViewSequences like ropes)
+
+impl<T> Default for AppendVec<T> {
+    fn default() -> Self {
+        Self {
+            inner: Default::default(),
         }
+    }
+}
 
-        impl<'a, 'b> $elements_splice for $crate::VecSplice<'a, 'b, $pod> {
-            fn push(&mut self, element: $pod, _cx: &mut $cx) {
-                self.push(element);
-            }
+/// Views for ordered sequences of elements.
+///
+/// Generally, a container view will internally contain a `ViewSequence`.
+/// The child elements of the container will be updated by the `ViewSequence`.
+///
+/// This is implemented for:
+///  - Any single [`View`], where the view's element type
+///    is [compatible](SuperElement) with the sequence's element type.
+///    This is the root implementation, by which the sequence actually
+///    updates the relevant element.
+///  - An `Option` of a `ViewSequence` value.
+///    The elements of the inner sequence will be inserted into the
+///    sequence if the value is `Some`, and removed once the value is `None`.
+///  - A [`Vec`] of `ViewSequence` values.
+///    Note that this will have persistent allocation with size proportional
+///    to the *longest* `Vec` which is ever provided in the View tree, as this
+///    uses a generational indexing scheme.
+///  - Tuples of `ViewSequences` with up to 15 elements.
+///    These can be nested if an ad-hoc sequence of more than 15 sequences is needed.
+///
+pub trait ViewSequence<State, Action, Context: ViewPathTracker, Element: ViewElement, Marker>:
+    'static
+{
+    type SeqState;
+
+    /// Build the associated widgets into `elements` and initialize all states.
+    #[must_use]
+    fn seq_build(&self, ctx: &mut Context, elements: &mut AppendVec<Element>) -> Self::SeqState;
```

Comment:
There are no previous elements (that this sequence should know about), so it's not meaningful for this to use the full splice.

Using AppendVec here simplifies management of this

Reply:
Originally using `ElementSplice` in `build` had the purpose of tracking the tree-structure (and in xilem_web adding children to the parent element in the DOM tree in place), and before that, it was just a `Vec`, but I guess when we're reimplementing tree-structure/xilem_web, we could revisit that. I want to investigate in xilem_web, to also use a separate element-tree (in Rust, not the DOM-tree I mean), so maybe it's not necessary, and we probably want to track the tree-structure directly in masonry anyway.

It was mostly a convenience thing, to reuse the trait, you're right, only elements are added in `build` (so IIRC only `ElementSplice::push` was needed).